internal/txapp/txapp.go

```diff
@@ -350,20 +350,61 @@ func (r *TxApp) Finalize(ctx context.Context, db sql.DB, block *common.BlockCont
 			Engine: r.Engine,
 		}, block)
 		if err != nil {
-			return nil, fmt.Errorf("error running end block hook: %w", err)
+			return nil, nil, nil, fmt.Errorf("error running end block hook: %w", err)
 		}
 	}
 
+	// Get the join requests approved by this node
+	approvedJoins, err = r.approvedJoinRequests(ctx, db)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
 	r.valMtx.Lock()
 	r.validators = finalValidators
 	r.valMtx.Unlock()
 
-	return finalValidators, nil
+	return finalValidators, approvedJoins, expiredJoins, nil
+}
+
+// approvedJoinRequests returns the validator join requests that have been approved by this node.
+func (r *TxApp) approvedJoinRequests(ctx context.Context, db sql.DB) ([]*types.Validator, error) {
+	readTx, err := db.BeginTx(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer readTx.Rollback(ctx)
+
+	// get all validator-join resolutions
+	activeJoins, err := voting.GetResolutionsByType(ctx, readTx, voting.ValidatorJoinEventType)
+	if err != nil {
+		return nil, err
+	}
+
+	var approvedJoins []*types.Validator
+	for _, join := range activeJoins {
+		var req voting.UpdatePowerRequest
+		if err := req.UnmarshalBinary(join.Body); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal join request: %w", err)
+		}
+
+		for _, voter := range join.Voters {
```

Comment:
Why do we query the db here? All we need to do is recognize that, if a tx is a vote tx for a validator approval AND that was made by the local node, then we should approve the validator.

The reason I have an issue with this is because we are doing a full table scan here at the end of every block, and the resolutions table could be quite large for networks that use oracles.